jax/_src/debugger/core.py

```diff
@@ -29,35 +28,79 @@
 from jax._src import util
 import numpy as np
 
+
+@tree_util.register_pytree_node_class
+class _DictWrapper:
+  keys: list[Any]
+  values: list[Any]
+
+  def __init__(self, keys, values):
+    self._keys = keys
+    self._values = values
+
+  def to_dict(self):
+    return dict(zip(self._keys, self._values))
+
+  def tree_flatten(self):
+    return self._values, self._keys
+
+  @classmethod
+  def tree_unflatten(cls, keys, values):
+    return _DictWrapper(keys, values)
+
+
+class _CantFlatten:
+  __repr__ = lambda _: "<cant_flatten>"
+cant_flatten = _CantFlatten()
+
+def _safe_flatten_dict(dct: dict[Any, Any]) -> tuple[list[Any],
+                                                     tree_util.PyTreeDef]:
+  # We avoid comparison between keys by just using the original order
+  keys, values = [], []
+  for key, value in dct.items():
+    try:
+      tree_util.tree_leaves(value)
+    except:
+      # If flattening fails, we substitute a sentinel object.
+      value = cant_flatten
+    keys.append(key)
+    values.append(value)
+  return tree_util.tree_flatten(_DictWrapper(keys, values))
+
+
 @tree_util.register_pytree_node_class
 @dataclasses.dataclass(frozen=True)
 class DebuggerFrame:
   """Encapsulates Python frame information."""
   filename: str
-  locals: Dict[str, Any]
-  globals: Dict[str, Any]
+  locals: Dict[Any, Any]
+  globals: Dict[Any, Any]
```

Comment:
I think these are always keyed by strings, but they might contain dicts which aren't keyed by strings, right?

Reply:
(Even for dicts which aren't keyed by strings, I think `Hashable` may be a better annotation than `Any`.)

Reply:
Done.